File: 10/prob-10.js
/**
 * @reference https://projecteuler.net/problem=10
 * @description The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.
 * @problem Find the sum of all the primes below two million.
 * @solution 142915828925
 * iojs --harmony_arrow_functions prob-10.js
*/

'use strict';

function isPrime(num) {
	const sqrt = Math.ceil(Math.sqrt(num) + 1);
	let div = 2;

	while ( div < sqrt ) {
		if ( num % div === 0 ) {
			return false;
		}
		div++;
	}
}

function getPrimes(limit) {
	let i = 3;
	let primes = [2];

	while ( primes[primes.length - 1] < limit ) {
		if ( isPrime(i) !== false ) {
			primes.push(i);
		}

		i++;
	}

	return getSum(primes);
}

function getSum(arr) {
	return arr.reduce((prev, curr) => {
		return prev + curr;
	});
}

console.log( getPrimes(2000000) );

File: 10/prob-10.rb
# @reference https://projecteuler.net/problem=7
# @description By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.
# @problem What is the 10,0001st prime number?
# @solution 142915828925
# ruby prob-7.rb

def isPrime(num)
	# man, this is so much terser than the js equivalent
	2.upto( Math.sqrt(num).to_i ) do |div|
		return false if num % div == 0
	end
end

def getPrimes(limit)
	i = 3
	primes = [2]

	while primes[primes.length - 1] < limit
		if isPrime(i) != false
			primes.push(i)
		end

		i += 1
	end

	return getSum(primes)
end

def getSum(arr)
	return arr.reduce(:+)
end

puts getPrimes(2000000)
